include/clock_control.h

```diff
@@ -23,6 +24,20 @@ extern "C" {
 /* Used to select all subsystem of a clock controller */
 #define CLOCK_CONTROL_SUBSYS_ALL	NULL
 
+typedef void (*clock_control_clbk_t)(struct device *dev, void *user_data);
+
+struct clock_control_user {
+	sys_dlist_t list;
+	clock_control_clbk_t clbk;
+	void *user_data;
+};
+
+enum clock_control_status {
```

Comment:
why not use errno values?

Reply:
There is no error status here so far.

Reply:
changed to return:
0 - running
-EALREADY - starting
-ESHUTDOWN - off

Reply:
I would have kept the original enum. errno are describing errors on calls, not a generic non-erroneous status.

Reply:
what about `clock_control_async_on` should that return 0 when running and `-EALREADY` or `-EINPROGRESS` when scheduled? It would also be good to clarify error code returned by `clock_control_on`. Currently it does not specify error codes and nordic driver returns `-EINPROGRESS` or 0. Others usually return 0 always.

Reply:
Indeed codes are missing.

In case of clock_control_on it is straigtforward: 
- 0, even if it is already on. It's simpler, no need to let the user handling an error which anyway means what he requested was a success (he wanted it on, it is on anyway).

In case of clock_control_async_on, since there is a callback you'll need to verify it (on the same subsystem):
- 0 if it went fine
- -EINPROGRESS if given cb/user_data are the same as running one
- -EBUSY is cb/user_data are different
- -ENOTSUP some clock controller are immediate, so such function won't be implemented for them. Am I doing the righs assumption here?

Reply:
> 0 if it went fine

do you mean that it's up and running or started? I'm still not sure if we should distinguish between started and starting state here. I see following use cases for distinction:
- User can use that information is to determine from which context callback was called.
- user can start clock in blocking manner `while (clock_control_async_on(dev, NULL, now)==-EINPROGRESS));`

`-EBUSY` i would rather return that if provided user_data is already in the list which is rather application error.

> -EINVAL if cb is not provided (seems mandatory to have one)

I don't think it has to be mandatory. This could be supported and once driver implements it then `counter_clock_on()` can be simply `counter_clock_async_on(dev, NULL, now)`.

Reply:
Good point about -EINPROGRESS, let's not distinguish started/running.
As well as for the -EINVAL removal, I did not see that possibility of clock_on pointing to async_on.

Reply:
enum is brought back. It contains `unknown` status which means lack of support for checking status.